src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs

```diff
@@ -13,140 +13,114 @@ namespace Microsoft.CodeAnalysis.Remote
         {
-            if (solution == null)
+            public NoOpClient(Workspace workspace) :
+                base(workspace)
+            {
+            }
+
+            public override Task<Connection> TryCreateConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
             {
-                return null;
+                return SpecializedTasks.Default<Connection>();
             }
 
-            Contract.ThrowIfFalse(solution.Workspace == _workspace);
+            protected override void OnStarted()
+            {
+                // do nothing
+            }
 
-            var service = _workspace.Services.GetService<ISolutionSynchronizationService>();
-            return await service.CreatePinnedRemotableDataScopeAsync(solution, cancellationToken).ConfigureAwait(false);
+            protected override void OnStopped()
+            {
+                // do nothing
+            }
         }
 
-        // TODO: make this to not exposed to caller. abstract all of these under Request and Response mechanism
-        public abstract class Session : IDisposable
+        public abstract class Connection : IDisposable
         {
-            protected readonly PinnedRemotableDataScope PinnedScopeOpt;
             protected readonly CancellationToken CancellationToken;
+            private PinnedRemotableDataScope _scope;
 
             private bool _disposed;
 
-            protected Session(PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+            protected Connection(CancellationToken cancellationToken)
             {
                 _disposed = false;
+                _scope = null;
 
-                PinnedScopeOpt = scope;
                 CancellationToken = cancellationToken;
             }
 
-            public abstract Task InvokeAsync(string targetName, params object[] arguments);
+            protected abstract Task OnRegisterPinnedRemotableDataScopeAsync(PinnedRemotableDataScope scope);
 
-            public abstract Task<T> InvokeAsync<T>(string targetName, params object[] arguments);
-
-            public abstract Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync);
+            public PinnedRemotableDataScope PinnedRemotableDataScope => _scope;
 
-            public abstract Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync);
-
-            public void AddAdditionalAssets(CustomAsset asset)
+            public virtual Task RegisterPinnedRemotableDataScopeAsync(PinnedRemotableDataScope scope)
             {
-                Contract.ThrowIfNull(PinnedScopeOpt);
-                PinnedScopeOpt.AddAdditionalAsset(asset, CancellationToken);
+                // make sure all thread can read the info
+                Interlocked.Exchange(ref _scope, scope);
```

Comment:
This is still not threadsafe.  The read in the property does no synchronization so it may always return stale data.

Note: please just use locks if you want shared mutable state across threads